components/ipfs/ipfs_utils.cc

```diff
@@ -460,47 +464,110 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
       cid, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
 }
 
-// gateway.io/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy ->
-// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
-// bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy.ipfs.gateway.io ->
-// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
-absl::optional<GURL> TranslateToCurrentGatewayUrl(const GURL& url) {
-  if (!url.is_valid() || !url.SchemeIsHTTPOrHTTPS()) {
-    return absl::nullopt;
+// https://github.com/ipfs/specs/blob/main/http-gateways/SUBDOMAIN_GATEWAY.md#host-request-header
+std::string DecodeSingleLabelForm(const std::string& input) {
+  if (input.find('.') != std::string::npos) {
+    return input;
   }
+  std::string result;
+  const char* chars = input.c_str();
+  size_t i = 0;
+  for (i = 0; i < input.size(); i++) {
+    if (chars[i] == '-' && (i < input.size() - 1) && chars[i + 1] == '-') {
+      result.push_back('-');
+      i++;
+    } else if (chars[i] == '-') {
+      result.push_back('.');
+    } else {
+      result.push_back(chars[i]);
+    }
+  }
+  return result;
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayHost(const GURL& url) {
   std::vector<std::string> host_parts = base::SplitStringUsingSubstr(
       url.host(), ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
 
-  if (host_parts.size() > 2 && IsValidCID(host_parts.at(0)) &&
-      host_parts.at(1) == "ipfs") {
-    GURL final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+  if (host_parts.size() <= 2) {
+    return absl::nullopt;
+  }
+  GURL final_url;
+  if (host_parts.at(1) == "ipfs" && IsValidCID(host_parts.at(0))) {
+    final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+  } else if (host_parts.at(1) == "ipns" && IsValidIPNSCID(host_parts.at(0))) {
+    final_url = GURL("ipns://" + host_parts.at(0) + url.path());
+  } else if (host_parts.at(1) == "ipns") {
+    std::string decoded = DecodeSingleLabelForm(host_parts.at(0));
+    final_url = GURL("https://" + decoded + url.path());
+  }
+
+  if (!final_url.is_valid()) {
+    return absl::nullopt;
+  }
+
     GURL::Replacements replacements;
     replacements.SetQueryStr(url.query_piece());
     replacements.SetRefStr(url.ref_piece());
     return final_url.ReplaceComponents(replacements);
-  }
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayPath(const GURL& url) {
```

Comment:
No unittest for this one?

Reply:
A comment here too like what we have before ExtractSourceFromGateway would be useful.